stdlib/internals.ncl

```diff
@@ -123,24 +123,36 @@
     else
       %blame% (%label_with_message% "not a record" label),
 
-  "$forall_tail" = fun sealing_key acc label value =>
+  "$forall_tail" = fun sealing_key constr acc label value =>
     let current_polarity = %polarity% label in
     let polarity = (%lookup_type_variable% sealing_key label).polarity in
+    let plural = fun list => if %length% list == 1 then "" else "s" in
     if polarity == current_polarity then
       if value == {} then
         let tagged_label = %label_with_message% "polymorphic tail mismatch" label in
         let tail = %record_unseal_tail% sealing_key tagged_label value in
         acc & tail
       else
         let extra_fields = %fields% value in
-        let plural = if %length% extra_fields == 1 then "" else "s" in
         %blame%
           (
             %label_with_message%
-              "extra field%{plural} `%{std.string.join ", " extra_fields}`"
+              "extra field%{plural extra_fields} `%{std.string.join ", " extra_fields}`"
               label
           )
     else
+      let conflicts = std.array.filter
+        (fun field => std.array.elem field constr)
+        (%fields% value)
```

Comment:
Using standard library functions in the internals module makes me a bit scared, because this can relatively easily lead to infinite recursions. Because most standard library functions carry type annotations, these types get converted to contract checks when they get called, which causes `subcontract` in the interpreter to get called. And if the constellation of types is slightly unlucky that ends up producing a recursion. I think in this case it would be fine, but `internals.ncl` is special in this way.

Even if it's ugly, we should inline the definitions of `filter` and `elem` here.

Reply:
This is a good point - I hadn't really considered this. There are 9 references to `std.` in this file. Do you think it's worth rewriting all of them?

Reply:
Except for the reference to `std.contract.Equal`, I think that would be a good idea.

Reply:
As discussed on Matrix ([thread](https://matrix.to/#/!CTCVXeJTucJbViJokz:matrix.org/$Cx3VUTcBEzOUIJQk2CQNddxsuYrcyya2wlm1xxkprQE?via=matrix.org&via=kleen.org&via=matrix.dapp.org.uk)), this is extra painful because `internals.ncl` _must_ currently be a syntactic record - i.e., it's not enough that it just evaluates to a record.

I proposed an alternative solution whereby we split standard library function implementations out into a module `std.unsafe`, and then re-export them in the existing modules with type & contract annotations provided on top. This means we can (safely) call `std.unsafe` functions from within `internals` without worrying about infinite recursion.

Reply:
I spiked the `std.unsafe` idea, and it's gonna be a bit more complex than I'd hoped (because, e.g., `std.unsafe.array.length` gets inferred as `Dyn` and then we can't unify that with `std.array.length: forall a. Array a -> Number`). It might be possible to get the sort of API we want using codegen to avoid duplicating definitions, but that's probably not something we should rush into.

As such, we have two options:
- Leave these calls as-is, and tackle every call to `std` in this file at the same time, or
- Fix these two callsites, and come back for the rest of the file once we've worked out a better solution.

I'm maybe leaning towards the first option, but I don't think I have strong feelings either way.

Reply:
Another solution to this problem could be to have a special type annotation that doesn't generate a corresponding contract at runtime. I guess those would need to be forbidden outside of the stdlib (enforced by the parser), but that would allow to:

- write a single implementation that is statically typechecked
- export it either without any contract nor type (`std.unsafe`) or wrapped inside a type annotation (the current  `std` interface)

I don't know if that would work so well with mutually recursive functions, but at least self-contained ones such as `length` could be handled that way.

Reply:
I made [this issue](https://github.com/tweag/nickel/issues/1334) for discussing this issue further. Can we resolve the thread on this PR and leave in `std.` for now?

Reply:
Since in this case we're not actually producing infinite recursions, I agree we should keep this as is in the interest of moving this PR forward 👍

Reply:
By the way, it's not only about infinite recursion, but for performances as well. Contract do have a non-trivial impact on performance, and this is also why we try to refrain from firing contracts from within contracts in a non obvious way. For example, I believe we're also using `%length%` instead of `array.length` although there's no risk of recursion. This can be reworked later, so let's not block this PR on this, but just to give a bit more context. `internals` are really low-level functions that are called implicitly, potentially quite a lot, by the interpreter, so we've been trying to keep them low level as well.